lib/config/config.js

```diff
@@ -145,6 +169,14 @@ class Config {
       this.logger = logging.createLogger(this.logLevel, customLogger)
     }
 
+    // Print logging preamble now that the logger is completelly resolved
+    const sources = {
+      environment: envOptions,
+      start: opts,
+      file: confFile || {}
+    }
+    printLoggingPreamble(this, sources, configFilePath)
```

Comment:
I suspect it will make this more difficult for processing, but I think the logging preamble should come *after* the fallback serviceName/serviceVersion block below and *after* the `normalize(...)` call. It should show the effective values for some fields. E.g.:

- If all of `envOptions`, `opts`, and `confFile` set a `serviceName`, but the winner is invalid, then the logging preamble should show `null` for the `serviceName`.
- If the serviceName isn't specified, but the default value is determined below, the logging preamble should show it.
- If an invalid value for some config var is shown, and `normalize(...)` ignores it, the logging preamble should **not** include that config var unless it is one of the "Log Always" vars.

I wonder if this could be printed in `Agent.prototype.start` ... before the `if (!this._conf.active) {` block.

(Aside: it would be nice to include `this._agentActivationMethod` and perhaps drop the existing `agent configured correctly` log message. However, that brings up other questions, like whether we drop the `startTrace`, `main` and `dependencies`. :)

Reply:
I think here there are two views of the configuration:
- the configuration given to the agent
- the configuration that finally is resolved

The spec is not clear about what we need to show here, on each case the value is different
- showing the given configuration users can inspect if correctly passed with the readable format for some options like durations (`2s`, ...)
- showing resolved configuration users can se in which final configuration the agent is working with.

Notice that with the current implementation `normalize()` already uses the logger and, although the spec does not say the preamble should be the very 1st message, it would look strange IMHO.

On a previous comment you shared a JSON structure for config
```
agentConfig: {
  "serviceName": {
    "value": "my-service",
    "source": "env",
    "normalizedName": "service_name"
  },
  ...
}
```

Is that `normalizedName` property the result of the normalization? Do you think we can show both provided and normalized value? If so, is it necessary to log warnings about configs?

Reply:
Thank for the extensive feedback. Following it and since the agent is also resolving some configurations I made some changes.
- config resolves a new property with information for the logging preamble
- agent gets that info at `start()` and enriches it
- agent becomes responsible to print the preamble through the logger



> if a config var is shadowed by a higher config level (e.g. ELASTIC_APM_SERVICE_NAME overrides a value in agent.start(opts)), then it might be nice to have a log.debug for this

I agree


> > Is that normalizedName property the result of the normalization?

> That's coming from the spec line that says the normalized snake_case name should be logged.
> However, I'm not sure if we need to do this, or need to for all config vars. It might be useful for some of them.

It was easier to normalize all so it's done for all configuration options.